internalshared/configutil/kms.go

```diff
@@ -68,51 +68,49 @@ func parseKMS(result *[]*KMS, list *ast.ObjectList, blockName string, maxKMS int
 			key = item.Keys[0].Token.Value().(string)
 		}
 
-		// We first decode into a map[string]interface{} because purpose isn't
-		// necessarily a string. Then we migrate everything else over to
-		// map[string]string and error if it doesn't work.
-		var m map[string]interface{}
-		if err := hcl.DecodeObject(&m, item.Val); err != nil {
-			return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
-		}
-
+		var disabled bool
 		var purpose []string
 		var err error
-		if v, ok := m["purpose"]; ok {
-			if purpose, err = parseutil.ParseCommaStringSlice(v); err != nil {
-				return multierror.Prefix(fmt.Errorf("unable to parse 'purpose' in kms type %q: %w", key, err), fmt.Sprintf("%s.%s:", blockName, key))
+		{
+			// We first decode into a map[string]interface{} because purpose isn't
+			// necessarily a string. Then we migrate everything else over to
+			// map[string]string and error if it doesn't work.
+			var m map[string]interface{}
+			if err := hcl.DecodeObject(&m, item.Val); err != nil {
+				return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
 			}
-			for i, p := range purpose {
-				purpose[i] = strings.ToLower(p)
+
+			if v, ok := m["purpose"]; ok {
+				if purpose, err = parseutil.ParseCommaStringSlice(v); err != nil {
+					return multierror.Prefix(fmt.Errorf("unable to parse 'purpose' in kms type %q: %w", key, err), fmt.Sprintf("%s.%s:", blockName, key))
+				}
+				for i, p := range purpose {
+					purpose[i] = strings.ToLower(p)
+				}
 			}
-			delete(m, "purpose")
-		}
 
-		var disabled bool
-		if v, ok := m["disabled"]; ok {
-			disabled, err = parseutil.ParseBool(v)
-			if err != nil {
-				return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
+			if v, ok := m["disabled"]; ok {
+				disabled, err = parseutil.ParseBool(v)
+				if err != nil {
+					return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
+				}
 			}
-			delete(m, "disabled")
 		}
 
-		strMap := make(map[string]string, len(m))
-		for k, v := range m {
-			if vs, ok := v.(string); ok {
-				strMap[k] = vs
-			} else {
-				return multierror.Prefix(fmt.Errorf("unable to parse 'purpose' in kms type %q: value could not be parsed as string", key), fmt.Sprintf("%s.%s:", blockName, key))
-			}
+		var cfg map[string]string
+		if err := hcl.DecodeObject(&cfg, item.Val); err != nil {
+			return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
 		}
+		delete(cfg, "purpose")
```

Comment:
Don't you want these two deletions to happen prior to the DecodeObject call a few lines above?

Reply:
Prior to DecodeObject cfg is empty.

Reply:
Oh, I didn't realize you were decoding the same item twice, I figured the second time around you'd decode the first 'm' val. So this is probably not an issue so long as hcl doesn't choke on decoding a string array into a string.

Reply:
The test I have seems to show that it doesn't.  Are there other scenarios you think I should be testing too?

Reply:
Nope, not at this time!